service/history/hsm/tree.go

```diff
@@ -213,29 +240,23 @@ func (n *Node) Path() []Key {
 	return append(n.Parent.Path(), n.Key)
 }
 
-// Outputs returns all outputs produced by transitions on this tree.
-func (n *Node) Outputs() []PathAndOutputs {
+// Outputs returns all operations that are relevant to this node and its subtree. For non-deleted nodes, this includes
+// their transitions and any deletions in their subtree. For deleted nodes, this includes their deletion operation and
+// any deletions of their descendants, but no transitions.
+func (n *Node) Outputs() OperationLog {
 	root := n.root()
 	currentPath := n.Path()
 
-	if root.opLog.IsDeleted(currentPath) {
-		return nil
-	}
-
-	var paos []PathAndOutputs
-	pathKey := fmt.Sprint(currentPath)
-	if transitions := root.opLog.TransitionsByPath[pathKey]; len(transitions) > 0 {
-		paos = append(paos, PathAndOutputs{
-			Path:    currentPath,
-			Outputs: transitions,
-		})
-	}
-
-	for _, child := range n.cache.children {
-		paos = append(paos, child.Outputs()...)
+	// Filter operations to only include those in this subtree
```

Comment:
You can also check if the current node is the root (e.g. parent is nil) and skip this loop since this how this be executed by the framework.

Reply:
Since the typical flow is for the root to be the arg, we now compact the full log and only filter if non-root. Thx.